sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala

```diff
@@ -1374,7 +1375,55 @@ case class RefreshTableCommand(tableIdent: TableIdentifier)
   override def run(sparkSession: SparkSession): Seq[Row] = {
     // Refresh the given table's metadata. If this table is cached as an InMemoryRelation,
     // drop the original cached version and make the new version cached lazily.
-    sparkSession.catalog.refreshTable(tableIdent.quotedString)
+    // If this table is a view, also refresh its underlying tables.
+    refreshAllUnderlyingTables(tableIdent, sparkSession)
     Seq.empty[Row]
   }
+
+  private def refreshAllUnderlyingTables(
+      tableIdent: TableIdentifier, sparkSession: SparkSession): Unit = {
+    val tableName = tableIdent.unquotedString
+    val catalog = sparkSession.sessionState.catalog
+    val table = catalog.getTempViewOrPermanentTableMetadata(tableIdent)
+    sparkSession.catalog.refreshTable(tableName)
+
+    if (table.tableType == CatalogTableType.VIEW) {
+      if (table.viewText.isDefined) {
+        val parser = sparkSession.sessionState.sqlParser
+        val unresolvedPlan = parser.parsePlan(table.viewText.get)
+        findOutUnderlyingTables(unresolvedPlan).foreach { t =>
+          refreshAllUnderlyingTables(t, sparkSession)
+        }
+      } else {
+        catalog.getTempView(tableName)
+          .orElse(catalog.getGlobalTempView(tableName))
+          .map(findOutUnderlyingTables).foreach { tables =>
```

Comment:
In branch-2.3, the logical plan for temporary view is analyzed logical plan.

In master  branch, it is unresolved logical plan.

So I also enable refresh underlying tables for temporary view.